dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/WebSocketTests.java

```diff
@@ -0,0 +1,125 @@
+ * in JSF 4.0 specification for <f:websocket> onerror="...”.
+ */
+@RunWith(FATRunner.class)
+public class WebSocketTests {
+
+    private static final String WEB_SOCKET_TEST_APP_NAME = "WebSocket";
+    protected static final Class<?> c = WebSocketTests.class;
+
+    @Rule
+    public TestName name = new TestName();
+
+    @Server("faces40_WebSocketServer")
+    public static LibertyServer server;
+
+    @BeforeClass
+    public static void setup() throws Exception {
+        ShrinkHelper.defaultDropinApp(server, WEB_SOCKET_TEST_APP_NAME + ".war",
+                                      "io.openliberty.org.apache.faces40.fat.websocket");
+
+        // Start the server and use the class name so we can find logs easily.
+        server.startServer(WebSocketTests.class.getSimpleName() + ".log");
+    }
+
+    @AfterClass
+    public static void tearDown() throws Exception {
+        // Stop the server
+        if (server != null && server.isStarted()) {
+            server.stopServer("SRVE0190E"); // SRVE0190E is due to ENABLE_WEBSOCKET_ENDPOINT being false for triggering onerror on a websocket.
+        }
+    }
+
+    @Before
+    public void setupPerTest() throws Exception {
+        server.setMarkToEndOfLog();
+    }
+
+    /**
+     * Test to ensure that the <f:websocket> onerror listener works properly.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testOnErrorWebsocket() throws Exception {
+        try (WebClient webClient = new WebClient()) {
+            webClient.getOptions().setThrowExceptionOnScriptError(false);
+
+            // Construct the URL for the test
+            String contextRoot = "WebSocket";
+            URL url = JSFUtils.createHttpUrl(server, contextRoot, "OnErrorWebSocketTest.jsf");
+
+            HtmlPage testOnErrorWebSocketPage = (HtmlPage) webClient.getPage(url);
+
+            // Log the page for debugging if necessary in the future.
+            Log.info(c, name.getMethodName(), testOnErrorWebSocketPage.asText());
+            Log.info(c, name.getMethodName(), testOnErrorWebSocketPage.asXml());
+
+            // Verify that the page contains the expected messages.
+            assertContains(testOnErrorWebSocketPage.asText(),
+                           "JSF 4.0 WebSocket - Test that onerror is invoked correctly.");
+
+            // Get the form that we are dealing with
+            HtmlForm form = testOnErrorWebSocketPage.getFormByName("form1");
+
+            // Get the button that opens the push connection
+            HtmlSubmitInput openButton = form.getInputByName("form1:openButton");
+
+            // Now click the open button. This should result in an error since 'jakarta.faces.ENABLE_WEBSOCKET_ENDPOINT' is set to 'false'
+            HtmlPage openPage = openButton.click();
+
+            assertTrue(JSFUtils.waitForPageResponse(openPage, "Called onerror listener"));
+            assertTrue(JSFUtils.waitForPageResponse(openPage, "Called onclose listener"));
```

Comment:
It's odd that `JSFUtils.waitForPageResponse` works while `WebClient#waitForBackgroundJavaScript` does not. 

Thanks for finding a workaround, at least.

Reply:
Did you happen to try `webClient.setAjaxController(new NicelyResynchronizingAjaxController());` with javascript wait?

Reply:
I did some experimenting with it, but had an intermittent failure (1 out of 9 runs failed).

Reply:
Best to keep JSFUtils.waitForPageResponse then. I'll mark this as resolved. Thanks!